src/main/java/com/spotify/docker/client/ApacheUnixSocket.java

```diff
@@ -47,75 +49,78 @@
  */
 public class ApacheUnixSocket extends Socket {
 
-  private final AFUNIXSocket inner;
+  private final UnixSocketChannel inner;
 
   private final Queue<SocketOptionSetter> optionsToSet = Queues.newArrayDeque();
 
   public ApacheUnixSocket() throws IOException {
-    this.inner = AFUNIXSocket.newInstance();
+    this.inner = UnixSocketChannel.open();
   }
 
   @Override
   public void connect(final SocketAddress endpoint) throws IOException {
-    inner.connect(endpoint);
-    setAllSocketOptions();
+    if (endpoint instanceof UnixSocketAddress) {
+      inner.connect((UnixSocketAddress) endpoint);
+      setAllSocketOptions();
+    }
   }
 
   @Override
   public void connect(final SocketAddress endpoint, final int timeout) throws IOException {
-    inner.connect(endpoint, timeout);
```

Comment:
@rohansingh Did we accidentally stop setting `timeout` with this PR change?

Reply:
Looks like it's definitely not being set on the initial connection. Looking at it now, we could certainly call a setSoTimeout(int timeout) assuming that would do the job as expected. In fact, looking in jnr-unixsocket, It seems I added support for this ( https://github.com/jnr/jnr-unixsocket/commit/7f3ee30ff762d7bad5a441e0461bf74d494880e2 ), presumably so that this class would be able to continue making those calls.

Despite the call not being done on creation, I'm pretty sure that from my testing of pull #916 , both TCP and UNIX sockets would stop working when the read timeout was exceeded so it seems they're respecting it now. Although I also recall times in the past when timeouts were not respected at all on the UNIX socket.

Is there test case / set of calls that show UNIX sockets still don't fully respect the timeout ?

Reply:
@davidxia Sorry it's way too long ago, I barely remember the context.

Reply:
@rgrunber I'm not really that familiar with jnr. I just saw that the `timeout` arg wasn't being used at all in this method. Does the timeout get magically set somewhere else?

Reply:
Looks like I'm confusing connection timeout and socket timeout. Connection timeout is the maximum amount of time to wait for a connection to be established with the daemon. Socket timeout is the maximum amount of inactivity between 2 packets.

The timeout passed in to the connect() call is for connection timeout (value was 5000 ms), and you're right, it wouldn't be set anywhere else. I wonder if we could just call setSoTimeout initially with that value and have the socket behave the same. Eventually there are other calls to setSoTimeout done by Apache HttpComponents itself based on the RequestConfig values set in DefaultDockerClient. All of those calls are either for a timeout of 0 (no timeout client) or 30000 (30s timeout client).

Reply:
Is this issue now moot with these changes? https://github.com/spotify/docker-client/pull/918

Reply:
It's definitely a step in the right direction. Assuming all tests are working as expected with that change, then yes, that commit should resolve the issue.